### PR TITLE
OAuth Google

### LINKED ISSUE #15: Welcome page
As a visitor / user / admin,
When I visit the welcome page (at the root path)
It has the site’s name
It has a link to visit the landmark index

### LINKED ISSUE #21: Logging out
As a logged-in user,
When I visit any page,
I see a button to log out.
When I click that button,
I am redirected to the welcome page ‘/’,
And I see a button to log in.

### LINKED ISSUE #31: Registering a new account  - Google OAuth
As a visitor,
When I visit ‘/’,
I see a button to register.
When I click that button,
I see a prompt to enter my email and password or buttons to authenticate with Google
When I click the button to authenticate with Google, 
I am redirected to the authentication page for that site.
When I enter my login details there,
I am redirected to a form to enter my additional details.
When I enter my additional details,
I am redirected to the landmark index,
I see a message “Welcome <my first_name>!”
And I see a button to log out.

### LINKED ISSUE #32: Logging in - Google OAuth
As a registered (but not logged-in) user with Google OAuth,
When I visit ‘/’,
I see a button to log in.
When I click that button,
I see a prompt to enter my email and password or buttons to log in with Google
When I click the button to login with Google, 
I am redirected to the login page for that site.
When I enter my login details there,
I am redirected to the landmark index,
I see a message “Welcome <my email>!”
And I see a button to log out.

### PR DESCRIPTION
**Pull request by chakeresa**
This branch addresses several user stories:
 - Adds a basic welcome page (resolves #15)
 - Adds registration & login through Google OAuth (resolves #31 & resolves #32, respectively). Uses the `omniauth-google-oauth2` gem.
 - Adds ability to log out (resolves #21)

This branch also makes the following major changes:
 - Adds `first_name` `last_name` `email` & `display_name` columns to `users`
 - Adds the `google_auths` & `app_auths` tables / models (which store `user_id` in both, as well as `username` & `password` in `app_auths, and `uid` & `token` in `google_auths`)
 - Adds the `GoogleAuthsController` & `AppAuthsController`.
 - Moves the `has_secure_password` from `User` to `AppAuth`
 - Adds a **rake task (`migrate_existing_user_logins`) to be run in the production environment after deployment / migration** which moves the existing users' data (username & password) from `users` to `app_auths`

This branch also fixes some formatting issues:
 - Fixes the mobile scaling issue (previously was showing a tiny version of the page)
 - Moves the location search field into the nav bar
 - Formats the login & register forms nicely

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -3,18 +3,19 @@ def new
   end
   
   def create
-    user = User.find_by(username: params[:username].downcase)
-    if user && user.authenticate(params[:password])
-      session[:user_id] = user.id
-      flash[:success] = "Welcome, #{user.username}!"
```
Comment (MillsProvosty) [inline app/controllers/sessions_controller.rb]: nice refactor

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -3,18 +3,19 @@ def new
...
-      redirect_to landmarks_path
+    app_auth = AppAuth.find_by(username: params[:username].downcase)
+    @user = app_auth.user if app_auth
+    if @user && app_auth.authenticate(params[:password])
+      successful_login
     else
-      flash[:danger] = 'Incorrect username/password combination'
+      flash.now[:danger] = 'Incorrect username/password combination'
```
Comment (MillsProvosty) [inline app/controllers/sessions_controller.rb]: .now!! Fun!

File: app/views/shared/_nav.html.erb [hunk 3]
```diff
@@ -20,7 +28,7 @@
           <a class="btn btn-primary my-2 my-sm-0" href="<%= login_path %>">Login</a>
         </li>
         <li class="nav-item">
-          <a class="btn btn-primary my-2 my-sm-0" href="<%= new_user_path %>">Register</a>
+          <a class="btn btn-primary my-2 my-sm-0" href="<%= register_path %>">Register</a>
```
Comment (MillsProvosty) [inline app/views/shared/_nav.html.erb]: Good clear re-naming on register path.

File: spec/models/google_auth_spec.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+RSpec.describe GoogleAuth, type: :model do
+  describe 'relationships' do
+    it { should belong_to(:user).optional }
```
Comment (lpile) [inline spec/models/google_auth_spec.rb]: Did not know `.optional` was a thing. Nice to know!

**Pull request merged**